[WPE] Test external module loading and WPE Display

GeorgesStavracas opened the pull request:
#### fc66aceff14cad31572506bddef34af5d7241e5c
<pre>
[WPE] Test external module loading and WPE Display
<a href="https://bugs.webkit.org/show_bug.cgi?id=286919">https://bugs.webkit.org/show_bug.cgi?id=286919</a>

Reviewed by Carlos Garcia Campos and Adrian Perez de Castro.

Add a mock WPE platform that is absolutely barebones, and just returns
stub values for everything. It doesn&apos;t create views nor any other
platform object.

The mock platform exercises WPE&apos;s external module loading, with a
dynamic type registered at module loading time. It also has a higher
priority than regular WPE modules.

The new TestDisplay file only has a single test, that checks if
wpe_display_get_default() is able to properly find the mock platform
and load it as expected. It checks the resulting type name.

Further tests can be added by making the mock platform configurable.

* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WPEPlatform/TestDisplay.cpp: Added.
(testLoadExternalDisplay):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/TestWebKitAPI/glib/WebKitGLib/TestMain.cpp:
(main):
* Tools/TestWebKitAPI/wpe/mock-platform/CMakeLists.txt: Added.
* Tools/TestWebKitAPI/wpe/mock-platform/WPEDisplayMock.cpp: Added.
(wpeDisplayMockDispose):
(wpeDisplayMockConnect):
(wpeDisplayMockCreateView):
(wpeDisplayMockCreateInputMethodContext):
(wpeDisplayMockGetEGLDisplay):
(wpeDisplayMockGetKeymap):
(wpeDisplayMockGetPreferredDMABufFormats):
(wpeDisplayMockGetNScreens):
(wpeDisplayMockGetScreen):
(wpeDisplayMockGetDRMDevice):
(wpeDisplayMockGetDRMRenderNode):
(wpeDisplayMockUseExplicitSync):
(wpe_display_mock_class_init):
(wpe_display_mock_class_finalize):
(wpe_display_mock_init):
(g_io_module_load):
(g_io_module_unload):
* Tools/TestWebKitAPI/wpe/mock-platform/WPEDisplayMock.h: Added.

Canonical link: <a href="https://commits.webkit.org/289925@main">https://commits.webkit.org/289925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b10b41e87dd8eefb529f1497d90473027aa1233

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88449 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7968 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42887 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93407 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39203 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90500 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8355 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16152 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68216 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25935 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91451 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6390 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79999 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48582 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6162 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34412 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38311 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76534 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35298 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95249 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15624 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77082 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15880 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76341 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18785 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20729 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19148 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8662 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15640 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15381 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18830 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17163 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos reviewed (approved):
This looks good to me, I'll leave the cmake changes for @aperezdc

aperezdc commented on Tools/TestWebKitAPI/glib/CMakeLists.txt:
```diff
@@ -143,6 +144,25 @@ if (ENABLE_WPE_PLATFORM)
+    set_target_properties(TestWPEDisplay PROPERTIES RUNTIME_OUTPUT_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/TestWebKitAPI/WPEPlatform)
+    if (COMPILER_IS_GCC_OR_CLANG)
+        WEBKIT_ADD_TARGET_CXX_FLAGS(TestWPEDisplay -Wno-unused-parameter)
+    endif ()
```
One may use `WEBKIT_ADD_TARGET_CXX_FLAGS()` without checking `COMPILER_IS_GC_OR_CLANG` beforehand. The function actually checks whether the compiler supports the given flags and will only add them if supported.

```suggestion
    WEBKIT_ADD_TARGET_CXX_FLAGS(TestWPEDisplay -Wno-unused-parameter)
```

GeorgesStavracas replied:
Done!

aperezdc commented on Tools/TestWebKitAPI/wpe/mock-platform/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+    C_VISIBILITY_PRESET hidden
+    POSITION_INDEPENDENT_CODE TRUE
+    VISIBILITY_INLINES_HIDDEN TRUE
+    INTERFACE_INCLUDE_DIRECTORIES "${CMAKE_CURRENT_SOURCE_DIR}"
```
Any reason to not list this below in the `target_include_directories()` command? Using the `INTERFACE_*` properties defines settings to be used by _other_ things that link against `WPEMockPlatform`&mdash;and there is nothing else that links to it, so I think this line should not be needed.

GeorgesStavracas replied:
Yup, just removing it seems to work

aperezdc reviewed (commented):
The CMake parts are fine, I just left a couple of minor comments.

GeorgesStavracas committed:
[WPE] Test external module loading and WPE Display
https://bugs.webkit.org/show_bug.cgi?id=286919

Reviewed by Carlos Garcia Campos and Adrian Perez de Castro.

Add a mock WPE platform that is absolutely barebones, and just returns
stub values for everything. It doesn't create views nor any other
platform object.

The mock platform exercises WPE's external module loading, with a
dynamic type registered at module loading time. It also has a higher
priority than regular WPE modules.

The new TestDisplay file only has a single test, that checks if
wpe_display_get_default() is able to properly find the mock platform
and load it as expected. It checks the resulting type name.

Further tests can be added by making the mock platform configurable.

* Tools/TestWebKitAPI/PlatformWPE.cmake:
* Tools/TestWebKitAPI/Tests/WPEPlatform/TestDisplay.cpp: Added.
(testLoadExternalDisplay):
(beforeAll):
(afterAll):
* Tools/TestWebKitAPI/glib/CMakeLists.txt:
* Tools/TestWebKitAPI/glib/WebKitGLib/TestMain.cpp:
(main):
* Tools/TestWebKitAPI/wpe/mock-platform/CMakeLists.txt: Added.
* Tools/TestWebKitAPI/wpe/mock-platform/WPEDisplayMock.cpp: Added.
(wpeDisplayMockDispose):
(wpeDisplayMockConnect):
(wpeDisplayMockCreateView):
(wpeDisplayMockCreateInputMethodContext):
(wpeDisplayMockGetEGLDisplay):
(wpeDisplayMockGetKeymap):
(wpeDisplayMockGetPreferredDMABufFormats):
(wpeDisplayMockGetNScreens):
(wpeDisplayMockGetScreen):
(wpeDisplayMockGetDRMDevice):
(wpeDisplayMockGetDRMRenderNode):
(wpeDisplayMockUseExplicitSync):
(wpe_display_mock_class_init):
(wpe_display_mock_class_finalize):
(wpe_display_mock_init):
(g_io_module_load):
(g_io_module_unload):
* Tools/TestWebKitAPI/wpe/mock-platform/WPEDisplayMock.h: Added.

Canonical link: https://commits.webkit.org/289925@main
